#5 (Ingame reference for pronouns and discord user name)

mart3323 opened the pull request:
# The plan
![image](https://user-images.githubusercontent.com/8430090/182029737-8cc786f5-c2f9-4cab-89ec-7da51b2ce836.png)

mart3323 committed:
feat(playerprofile): Pronouns and discord name in HoverText

mart3323 commented:
~~Hover effects require the use of TextComponents, which aren't in the spigot API jar.~~

~~Blocked by #9~~
✅ #9 has been completed, work on this may resume

mart3323 commented:
@JochemKuijpers Turns out that the Gson parser fails after using the /reload command (`zip file closed` errors, i think trying to look up some class in an older version of.., something)

Do we still want to have the messages configurable despite it breaking after reloads?
```yaml
nameplate: >-
  {
    "text": "{{PLAYER_NAME}}",
    "hoverEvent": {
      "action": "show_text",
      "contents": "§dPronouns: §f{{PRONOUNS}}\n§aDiscord: §f{{DISCORD}}"
    }
  }
```
As the alternative, i could build up the components in code - allowing only the contents of the nameplate to be configured. Hopefully building the components is fine and it's just the dynamic class lookup of the GSON parser that's causing errors

JochemKuijpers commented:
I think it's fine to have the JSON structure hard-coded and only its contents dynamic.

JochemKuijpers commented on src/main/java/playfriends/mc/plugin/features/playerprofile/SetDiscordEvent.java:
```diff
@@ -0,0 +1,34 @@
+import org.bukkit.event.HandlerList;
+import playfriends.mc.plugin.api.PlayerEvent;
+
+public class SetDiscordEvent extends Event {
```
Can extend `playfriends.mc.plugin.api.PlayerEvent` instead of `Event`.

mart3323 replied:
Should it? (i'm assuming yes now that you've commented on it)

I saw that and thought it would be weird to create a whole inheritance chain where each subclass adds a single extra property.
But in hindsight, that might be because i've been doing a lot of typescript recently, and types there are structural and don't care about inheritance, making readability the only concern

Will change

JochemKuijpers replied:
My reasoning for it is twofold:
- I like the indication that this is a `PlayerEvent`. Spigot seems to have a fairly large event hierarchy and actually already has a `PlayerEvent`, but sadly it's not public so I just created a copy here.
- It's a convenience to not have to duplicate the fields/getters (and avoid inconsistency with getter or fieldnames). There's negligable overhead to doing this vs duplicating code (though I admit it's not much 🤷), as all Java's methods are virtual anyway.

JochemKuijpers commented on src/main/java/playfriends/mc/plugin/features/playerprofile/SetPronounsEvent.java:
```diff
@@ -0,0 +1,34 @@
+import org.bukkit.event.HandlerList;
+import playfriends.mc.plugin.api.PlayerEvent;
+
+public class SetPronounsEvent extends Event {
```
Can extend `playfriends.mc.plugin.api.PlayerEvent` instead of `Event`.

JochemKuijpers commented on src/main/resources/config.yml:
```diff
@@ -53,6 +53,19 @@ afk:
 
+alias:
+  greeting:
+    no-pronouns:  "§7You have not set any pronouns. See §e/help pronouns"
```
We now have a mix of `§` and `&` color codes. I'd prefer to stick with `&` since it's easier to type on most keyboards.

mart3323 replied:
I personally don't like & because it can be ambiguous, for example in the phrase "get me a bow&arrow", but i will change it back for consistency

JochemKuijpers replied:
It seems to be the convention for Spigot plugins as far as I could see. I don't really mind either way, but I do like it being consistent.

JochemKuijpers replied:
(by which I mean, if you want to convert everything to simdollars, that's fine by me too 😃)

JochemKuijpers commented on src/main/java/playfriends/mc/plugin/features/playerprofile/PlayerProfileHandler.java:
```diff
@@ -0,0 +1,97 @@
+    private String pronounsSetMessage;
+
+    public PlayerProfileHandler(PlayerDataManager playerDataManager) {
+        System.out.println("PlayerProfileChat created");
```
If you want to log something, especially in the async chat method, you should use the plugin's Logger (you can pass it as a constructor argument) because of thread-safety with various buffers involved. And otherwise you'll bypass writing to the log file too.

mart3323 replied:
Will remove since it was for debugging 

(the event wouldn't fire because i was forgetting to recompile the jar after making changes)

JochemKuijpers commented on src/main/java/playfriends/mc/plugin/playerdata/PlayerData.java:
```diff
@@ -29,6 +29,14 @@ public class PlayerData {
     @Persistent("afk-enabled")
     private boolean isAfkEnabled = true;
 
+    /** The player's preferred pronouns. */
```
nitpick: I'd drop the word 'preferred' here, since you're talking about someone's actual pronouns and not simply a preference 🙂

mart3323 replied:
Oh my.., yes of course, will change that

Must've recalled the phrase from somewhere without really thinking about it

mart3323 commented:
Took care of issues from review, refactored, added override for /list command (so you could check someone's nameplate without having to wait for them to speak first).

For threads, would you prefer i resolve them when i *think* they're handled, or to leave them to the reviewer to resolve? (We do the latter at work, no idea what standard practice is)

mart3323 commented:
~~Missed a few section symbols (think i accidentally readded them doing the rebase + squash) - will update in a minute~~
Done ✅

JochemKuijpers commented:
Looks good to me 👍
